ydb/core/tx/schemeshard/schemeshard__login.cpp

```diff
@@ -67,37 +69,133 @@ struct TSchemeShard::TTxLogin : TSchemeShard::TRwTxBase {
+        if (!loginRequest.ExternalAuth && !AppData(ctx)->AuthConfig.GetEnableLoginAuthentication()) {
+            Result->Record.SetError("Login authentication is disabled");
+            return true;
+        }
+        if (loginRequest.ExternalAuth) {
+            return HandleExternalAuth(loginRequest);
+        }
+        return HandleLoginAuth(loginRequest, db, ctx);
+    }
+
+    bool HandleExternalAuth(const NLogin::TLoginProvider::TLoginUserRequest& loginRequest) {
+        const NLogin::TLoginProvider::TLoginUserResponse loginResponse = Self->LoginProvider.LoginUser(loginRequest);
+        switch (loginResponse.Status) {
+        case NLogin::TLoginProvider::TLoginUserResponse::EStatus::SUCCESS: {
+            SetSuccessResult(loginResponse);
+            break;
+        }
+        case NLogin::TLoginProvider::TLoginUserResponse::EStatus::INVALID_PASSWORD:
+        case NLogin::TLoginProvider::TLoginUserResponse::EStatus::INVALID_USER:
+        case NLogin::TLoginProvider::TLoginUserResponse::EStatus::UNAVAILABLE_KEY:
+        case NLogin::TLoginProvider::TLoginUserResponse::EStatus::UNSPECIFIED: {
+            Result->Record.SetError(loginResponse.Error);
+            break;
+        }
+        }
+        return true;
+    }
+
+    void SetSuccessResult(const NLogin::TLoginProvider::TLoginUserResponse& loginResponse) {
+        Result->Record.SetToken(loginResponse.Token);
+        Result->Record.SetSanitizedToken(loginResponse.SanitizedToken);
+    }
+
+    bool HandleLoginAuth(const NLogin::TLoginProvider::TLoginUserRequest& loginRequest, NIceDb::TNiceDb& db, const TActorContext& ctx) {
+        auto row = db.Table<Schema::LoginSids>().Key(loginRequest.User).Select();
+        if (!row.IsReady()) {
+            return false;
+        }
+        if (!row.IsValid()) {
+            Result->Record.SetError(TStringBuilder() << "Cannot find user: " << loginRequest.User);
+            return true;
+        }
+        CurrentFailedAttemptCount = row.GetValueOrDefault<Schema::LoginSids::FailedAttemptCount>();
+        TInstant lastFailedAttempt = TInstant::FromValue(row.GetValue<Schema::LoginSids::LastFailedAttempt>());
+        if (CheckAccountLockout()) {
+            if (ShouldUnlockAccount(lastFailedAttempt)) {
+                UnlockAccount(loginRequest, db);
+            } else {
+                Result->Record.SetError(TStringBuilder() << "User " << loginRequest.User << " is locked out");
+                return true;
             }
+        } else if (ShouldResetFailedAttemptCount(lastFailedAttempt)) {
+            ResetFailedAttemptCount(loginRequest, db);
+        }
+        const NLogin::TLoginProvider::TLoginUserResponse loginResponse = Self->LoginProvider.LoginUser(loginRequest);
+        switch (loginResponse.Status) {
+        case NLogin::TLoginProvider::TLoginUserResponse::EStatus::SUCCESS: {
+            HandleLoginAuthSuccess(loginRequest, loginResponse, db);
```

Comment:
`HandleLoginAuthSuccess()` and `HandleLoginAuthInvalidPassword()` both hide how the final `Result` is set. While the whole point of `HandleLoginAuth()` is to construct that `Result`. So hiding it does not help the reader to understand how `Result` is constructed.

I would suggest moving `Result` setting out of those functions and back to `HandleLoginAuth()`.

Reply:
done